### ISSUE TITLE
unserializing objects with dependencies

### ISSUE THREAD

**Issue by flip111** [symfony/symfony#12222]
Hi,

I was thinking about creating a Form Builder UI. The user could specify the fields and then a `FormBuilder` class would be created at runtime. Great !
Just then i noticed the FormBuilder can be feeded some classes that can be services too, like the `EventDispatcher`. So in case you need that one you need the `EventDispatcher` which was already instantiated by the DIC and not a new one.

After looking around a litlte bit the following resource showed up:
https://www.drupal.org/node/2004282
where @stof points to
https://github.com/symfony/symfony/pull/7230

I would like to know if it's feasible to add support for this to the DI component or if it's left to the developer to request services for the unserialized component and figure out how to inject them.

There are some problems with doing it this way. Please consider the general question of injected services into unserialized objects, not just in the context of the FormBuilder, but also as in the linked PR of assetic and other use cases.

1. All the required services already need to be instantiated while they might not be needed depending on the type of request (a lazy proxy service already helps here).
2. Depending on business logic different services might need to be injected that can change in runtime.

Code could look like this:

```yml
# app/config/config.yml
services:
    my_service:
        class:        Some\Sort\OfClass
        arguments:    [@dependency1]
        calls:
            - [setDependency2, ["@dependency2"]]
```

```php
<?php
// Some sort of class
class OfClass {
  protected $dependency1;
  protected $dependency2;
  protected $dependency3; // optional

  public function __construct($dependency1) {

  }

  public function setDependency2($dependency2) {

  }
}

// Somewhere else in another file
$fb = $container->wakeup(
  "some serialized FormBuilder string",
  [
    'dependency1' => $runtime_service1, // overwrite
    // dependency 2 --> default dependency
    'dependency3' => $runtime_service3 // supplemented
  ]
);
```

DIC would then do the following:
1. unserialize the object
2. instantiate dependency graph when not available (optional step i guess)
3. choose different dependencies if specified in the `wakeup()` method
4. inject dependencies. Possible implementation: reflection (a la Doctrine)
5. give back object at runtime.

In addition to my question if this would be a good suggestion to the DIC component in specific, i would like to ask if this would be a good attribute to the symfony code wether or not it will be placed in the DIC component.

In conclusion: it's not that hard to make the *ugly* way when all the services are available from the DIC, but it might be worthwhile to have this functionality available to everyone.

Thoughts?

**Comment by fabpot** [comment]
Closing.

**Comment by flip111** [comment]
In my opinion the [introduction to the service container](http://symfony.com/doc/current/book/service_container.html) is ambigious.

`Put simply, a Service is any PHP object that performs some sort of "global" task.`
global seems ambigious because it's in quotes. This indicates that there are also other tasks (non-global).

Then to continue:
`The advantage of thinking about "services" is that you begin to think about separating each piece of functionality in your application into a series of services.`
Together with the previous sentence that would imply that all your functionality in your application are global tasks. The rest of the application are don't have functionality, like Data-Transfer-Objects.

By all means if anyone thinks i'm pulling things out of context from each other, please elaborate on the correct context of how this is supposed to read.

The documentation points to a wikipedia page that says:
`A service-oriented architecture (SOA) is a design pattern in which application components provide services to other components via a communications protocol, typically over a network.`

Two things stand out here in my opinion. First that it talks about `components`. And i ask myself would any service class already classify as component?

Second that typically this kind of architecture is used over network (as supposed to classes "communicating" to each other on the same machine). For me the wikipedia page relates more to for example the trend of micro-services that are becomming more popular lately.

I saw people in #symonfy @ freenode put every piece of functionality in a service class and prevents the data-objects from modifying themselves. Some sources say this is bad practice ["(OOP) Writing lots of "xxxxxManager" classes that contain all of the methods for manipulating the fields of objects that have little or no methods of their own"](http://www.yacoset.com/Home/signs-that-you-re-a-bad-programmer)

I'm going a stretch on my intuition here. What i would like to know the following:

* Is it desired that symfony dictate the architecture of application specific code? (of course it must dictate it's own architecture!)
* Each architecture brings it's own tradeoffs. There are reasons mentioned why to promote the service orientated architecture such as `Each service can also be more easily tested and configured since it's separated from the other functionality in your application.`. `More easily tested` than opposed to???

I realise any discussion on architecture would be more difficult then let's say an implementation detail of a method. I'm sad the issue was closed so early as i don't see many issues covering the bigger picture of the symfony framework...which is also important for a community driven framework, no?

**Comment by xabbuh** [comment]
@flip111 Do you mind opening an issue on the docs repository if you think that the documentation should be improved/clarified in a way?
